### PR TITLE
refactor(frontend): Use Zod for TokenId

### PR DESCRIPTION
**Pull request by AntonioVentilii-DFINITY**
# Motivation

We want to use Zod to safe parse the types of Token. So, we start doing it for type TokenId, branding it specifically for TokenId.

# Changes

- Create schema for `TokenId`.
- Use it to define `TokenId` type.
- Create util to parse `TokenId` type.
- Adapt the code in each usage of `TokenId`.

# Tests

I didn't create a test for the util because I am not sure how to make it throw errors, but open to suggestions.

### CODE REVIEW

**Review by peterpeterparker — COMMENTED** [review]
Looks good!

- Instead of `zod.utils.ts`, should we consider creating a new package to hold these "type utilities and Zod parsers" and naming the modules with "tokens" something?

- Should we add some tests for the Zod parsers, or that's too much?

File: scripts/add.tokens.erc20.mjs [hunk 1]
```diff
@@ -181,7 +183,7 @@ import ${icon} from '$icp-eth/assets/${icon}.svg';
 			? `
 export const ${symbol}_SYMBOL = '${mainnetSymbol}';
 
-export const ${symbol}_TOKEN_ID: unique symbol = Symbol(${symbol}_SYMBOL);
+export const ${symbol}_TOKEN_ID: TokenId = parseTokenId(${symbol}_SYMBOL);
```
[the remark this answers is not in the document]
Comment (AntonioVentilii-DFINITY) [reply scripts/add.tokens.erc20.mjs:186]: that is only the string that is going to be written in the TS file created by the script: the script created a new TS file for each new token, and it needs to set all the variables. I am adjusting to the new way of defining `TokenId`

**Comment by AntonioVentilii-DFINITY** [comment]
@peterpeterparker 

- You mean a complete new folder that is a peer of `utils` ? Something like `lib/parser` or `lib/zod` ?

- For sure, but for this specific one, I could not bring myself to find a good "error" test. Theoretically, it will always work if the input is type-sefe (a string):

describe("parseTokenId", () => {
  it("should parse a valid token ID string into a TokenId", () => {
    const result = parseTokenId('token-id');
    expect(result).toBeInstanceOf(Symbol);
    expect(result.toString()).toBe(`Symbol(token-id})`);
  });

  it("should throw a ZodError for an invalid token ID string", () => {
    const invalidTokenIdString =  ???

    expect(() => parseTokenId(invalidTokenIdString)).toThrow();
  });
  ...

File: scripts/add.tokens.erc20.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (peterpeterparker) [reply scripts/add.tokens.erc20.mjs:186]: Not sure what you mean. What I mean is that `: unique symbol` in a `mjs` file was already suprising.

```
// Not ok
export const ${symbol}_TOKEN_ID: unique symbol;

// ok
export const ${symbol}_TOKEN_ID;
```

Or the script should not be a `mjs` file but a `ts` file.

**Comment by peterpeterparker** [comment]
> You mean a complete new folder that is a peer of utils ? Something like lib/parser or lib/zod ?

Yes. I like `lib/parser`. Or maybe `lib/schema` or `lib/validation` or `lib/schema-validation` etc. I was thinking its maybe not a utils?

> For sure, but for this specific one, I could not bring myself to find a good "error" test. Theoretically, it will always work if the input is type-safe (a string):

Good point. We can test the positive case only here. I was more asking generally speaking.

**Comment by AntonioVentilii-DFINITY** [comment]
> or `lib/validation `

going with this! Shall it contain the schema too? or we leave that in the `type` folder?


> I was more asking generally speaking.

Yes, completely agree, I will create tests for the next parsers

**Comment by peterpeterparker** [comment]
> Shall it contain the schema too? or we leave that in the type folder?

As you wish, no particular opinion.

>  I will create tests for the next parsers

👍

File: scripts/add.tokens.erc20.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (AntonioVentilii-DFINITY) [reply scripts/add.tokens.erc20.mjs:186]: Nono, wait, what i meant is that that part is not part of the code, it is a string,  a big string, so it has no effect, nor it should be considered as code:

```javascript
const newFileContentTestnet =
		nonNullish(testnetToken) && !testnetTokenCreated
			? `
export const SEPOLIA_${symbol}_SYMBOL = 'Sepolia${testnetSymbol}';

export const SEPOLIA_${symbol}_TOKEN_ID: TokenId = parseTokenId(SEPOLIA_${symbol}_SYMBOL);

export const ${testnetToken}: RequiredErc20Token = {
	id: SEPOLIA_${symbol}_TOKEN_ID,
	network: SEPOLIA_NETWORK,
	standard: 'erc20',
	category: 'default',
	name: '${testnetName}',
	symbol: '${testnetSymbol}',
	decimals: ${symbol}_DECIMALS,
	icon: ${icon},
	address: '${testnetContractAddress}',
	exchange: 'erc20',
	twinTokenSymbol: 'ckSepolia${symbol}'
};
`
			: '';
```

[the remark this answers is not in the document]
Comment (peterpeterparker) [reply scripts/add.tokens.erc20.mjs:186]: Apologies.

**Review by peterpeterparker — APPROVED** [review]
LGTm, thx

**Pull request merged**